src/main/java/com/squareup/javapoet/Util.java

```diff
@@ -32,6 +38,72 @@
   private Util() {
   }
 
+  private static final Modifier DEFAULT;
+  /**
+   * Used to detect conflicting modifiers, maps from a modifier to the
+   * conflicting modifier list.
+   */
+  private static final Map<Modifier, Set<Modifier>> CONFLICTING_MODIFIER_MAP;
+  static {
+    Modifier def = null;
+    try {
+      def = Modifier.valueOf("DEFAULT");
+    } catch (IllegalArgumentException expected) {
+    }
+    DEFAULT = def;
+    Map<Modifier, Set<Modifier>> temp = new EnumMap<Modifier, Set<Modifier>>(Modifier.class) {
+      private static final long serialVersionUID = 879880137764123586L;
+
+      @Override public Set<Modifier> put(Modifier key, Set<Modifier> value) {
+        // This handles name collisions
+        if (containsKey(key)) {
+          Set<Modifier> val = get(key);
+          val.addAll(value);
+          return val;
+        }
+        return super.put(key, value);
+      }
+
+      @Override public void putAll(
+          Map<? extends Modifier, ? extends Set<Modifier>> m) {
+        for (Map.Entry<? extends Modifier, ? extends Set<Modifier>> entry : m.entrySet()) {
+          put(entry.getKey(), entry.getValue());
+        }
+      }
+    };
+    temp.putAll(generateCombinations(Modifier.PUBLIC, Modifier.PROTECTED, Modifier.PRIVATE));
+    temp.putAll(generateCombinations(Modifier.ABSTRACT, DEFAULT, Modifier.PRIVATE));
+    temp.putAll(generateCombinations(Modifier.ABSTRACT, DEFAULT, Modifier.STATIC));
+    temp.putAll(generateCombinations(Modifier.ABSTRACT, Modifier.STRICTFP));
+    temp.putAll(generateCombinations(Modifier.ABSTRACT, Modifier.PRIVATE));
+    temp.putAll(generateCombinations(Modifier.ABSTRACT, Modifier.STATIC));
+    temp.putAll(generateCombinations(Modifier.ABSTRACT, Modifier.FINAL));
+    temp.putAll(generateCombinations(Modifier.ABSTRACT, Modifier.NATIVE));
+    temp.putAll(generateCombinations(Modifier.ABSTRACT, Modifier.SYNCHRONIZED));
+    temp.putAll(generateCombinations(Modifier.NATIVE, Modifier.STRICTFP));
+
+    EnumMap<Modifier, Set<Modifier>> result = new EnumMap<>(Modifier.class);
+    for (Map.Entry<Modifier, Set<Modifier>> entry : temp.entrySet()) {
+      if (entry.getValue().isEmpty()) continue;
+      result.put(entry.getKey(), immutableSet(entry.getValue()));
+    }
+    CONFLICTING_MODIFIER_MAP = Collections.unmodifiableMap(result);
+  }
+
+  private static Map<Modifier, Set<Modifier>> generateCombinations(Modifier... source) {
+    Set<Modifier> sourceNonNull = new HashSet<>(Arrays.asList(source));
```

Comment:
nit: never, never, never use HashSet. Always use LinkedHashSet. The extra determinism is worth it.